presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java

```diff
@@ -0,0 +1,323 @@
+                    resultKeySet.add(
+                            other.mapBlock != null ? other.mapBlock : other.keySet.getBlock(),
+                            other.mapBlock != null ? 2 * i : i);
+                    other.appendValue(i, resultValueBlockBuilder);
+                }
+            }
+        }
+
+        assert resultKeySet.size() == resultValueBlockBuilder.getPositionCount();
+        return new KeySetAndValues(keyType, valueType, resultKeySet, resultValueBlockBuilder.build(), null);
+    }
+
+    public KeySetAndValues unionSum(Block value)
+    {
+        KeySetAndValues keySetAndValues = new KeySetAndValues(keyType, valueType, null, null, value);
+        return unionSum(keySetAndValues);
+    }
+
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE + (mapBlock != null ? 0 : keySet.getRetainedSizeInBytes() + valueBlock.getRetainedSizeInBytes());
+    }
+
+    private static final Adder DOUBLE_ADDER = new Adder() {
+        @Override
+        public void writeSum(Type type, Block block1, int position1, Block block2, int position2, BlockBuilder blockBuilder)
+        {
+            type.writeDouble(blockBuilder, longBitsToDouble(block1.getLong(position1)) + longBitsToDouble(block2.getLong(position2)));
+        }
+    };
+
+    private static final Adder FLOAT_ADDER = new Adder() {
+        @Override
+        public void writeSum(Type type, Block block1, int position1, Block block2, int position2, BlockBuilder blockBuilder)
+        {
+            type.writeLong(blockBuilder, Float.floatToIntBits(intBitsToFloat(block1.getInt(position1)) + intBitsToFloat(block2.getInt(position2))));
+        }
+    };
+
+    private static final Adder INT_ADDER = new Adder() {
+        @Override
+        public void writeSum(Type type, Block block1, int position1, Block block2, int position2, BlockBuilder blockBuilder)
+        {
+            type.writeLong(blockBuilder, block1.getInt(position1) + block2.getInt(position2));
+        }
+    };
+
+    private static final Adder LONG_ADDER = new Adder() {
+        @Override
+        public void writeSum(Type type, Block block1, int position1, Block block2, int position2, BlockBuilder blockBuilder)
+        {
+            type.writeLong(blockBuilder, block1.getLong(position1) + block2.getLong(position2));
+        }
+    };
+
+    private interface Adder
+    {
+        // Sum the two non-null values - value1[position1] and value2[position2]
+        void writeSum(Type type, Block block1, int position1, Block block2, int position2, BlockBuilder blockBuilder);
+    }
+
+    private static Adder getAdder(Type type)
+    {
+        if (BIGINT.equals(type)) {
+            return LONG_ADDER;
+        }
+
+        if (DOUBLE.equals(type)) {
+            return DOUBLE_ADDER;
+        }
+
+        if (REAL.equals(type)) {
+            return FLOAT_ADDER;
+        }
+
+        if (INTEGER.equals(type)) {
+            return INT_ADDER;
+        }
+
+        assert (false);
```

Comment:
What about TINYINT and SMALLINT? At least change this to throw proper internal errors. This looks very uncivilized. 😂

Reply:
the main issue is it's bug-prone to sum tinyint due to overflows and stuff. So I wasn't sure if that's actually worth implementing. But I can add it.

Reply:
also, we seem to be upconverting everything to integer? That's why I don't think it's even needed.

Reply:
Done